hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -0,0 +1,294 @@
+            else return (isPing msg, messageAckForParty, knownAckForParty, knownAcks)
+
+        if shouldCallback
+          then do
+            callback (Authenticated msg party)
+            traceWith tracer (Received acks knownAcks partyIndex)
+          else traceWith tracer (Ignored acks knownAcks partyIndex)
+
+        -- resend messages if party did not acknowledge our latest idx
+        resendMessages resend partyIndex sentMessages knownAcks acks messageAckForParty knownAckForParty
+
+        -- Update last message index sent by us and seen by some party
+        updateSeenMessages seenMessages acks party
+        -- Take the lowest number from seen messages by everyone and remove it from
+        -- our sent messages.
+        deleteSeenMessages sentMessages seenMessages
+
+  ignoreMalformedMessages = pure ()
+
+  constructAcks acks wantedIndex =
+    zipWith (\ack i -> if i == wantedIndex then ack + 1 else ack) acks partyIndexes
+
+  partyIndexes = generate (length allParties) id
+
+  resendMessages resend partyIndex sentMessages knownAcks messageAcks messageAckForParty knownAckForParty = do
+    myIndex <- findPartyIndex me
+    let messageAckForUs = messageAcks ! myIndex
+    let knownAckForUs = knownAcks ! myIndex
+    when (messageAckForUs < knownAckForUs && messageAckForParty <= knownAckForParty) $ do
+      let missing = fromList [messageAckForUs + 1 .. knownAckForUs]
+      messages <- readTVarIO sentMessages
+      forM_ missing $ \idx -> do
+        case messages IMap.!? idx of
+          Nothing ->
+            throwIO $
+              ReliabilityFailedToFindMsg $
+                "FIXME: this should never happen, there's no sent message at index "
+                  <> show idx
+                  <> ", messages length = "
+                  <> show (IMap.size messages)
+                  <> ", latest message ack: "
+                  <> show knownAckForUs
+                  <> ", acked: "
+                  <> show messageAckForUs
+          Just missingMsg -> do
+            let newAcks' = zipWith (\ack i -> if i == myIndex then idx else ack) knownAcks partyIndexes
+            traceWith tracer (Resending missing messageAcks newAcks' partyIndex)
+            atomically $ resend $ ReliableMsg newAcks' missingMsg
+
+  updateSeenMessages seenMessages acks party = do
+    myIndex <- findPartyIndex me
+    let messageAckForUs = acks ! myIndex
+    atomically $ modifyTVar' seenMessages (Map.insert party messageAckForUs)
+
+  deleteSeenMessages sentMessages seenMessages = do
+    clearedMessages <- atomically $ do
+      seenMessages' <- readTVar seenMessages
+      let messageReceivedByEveryone =
+            case sortBy (\(_, a) (_, b) -> compare a b) (Map.toList seenMessages') of
+              [] -> 0 -- should not happen
+              ((_, v) : _) -> v
+      sentMessages' <- readTVar sentMessages
+      if IMap.member messageReceivedByEveryone sentMessages'
```

Comment:
I do not understand how the removal of messages work. The function name indicates it deletes multiple messages, but the type of `messageReceivedByEveryone` is just a single `Key` (`Int`). Maybe adding some comments would be justified here?